docs/VERSIONING.rst

```diff
@@ -0,0 +1,694 @@
+
+Creation of New Versions
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+When versioning is enabled in a bucket, APIs which normally result in
+the creation of objects, such as Put Object, Complete Multipart Upload
+and Copy Object, will generate new versions of objects.
+
+S3 creates a new version and updates the master version using the
+``versioning: true`` option in PUT calls to the metadata engine. As an
+example, when two consecutive Put Object requests are sent to the S3
+Server for a versioning-enabled bucket with the same key names, there
+are two corresponding metadata PUT calls with the ``versioning`` option
+set to true.
+
+The PUT calls to metadata and resulting keys are shown below:
+
+(1) PUT foo (first put), versioning: ``true``
+
++----------+---------+
+| key      | value   |
++==========+=========+
+| foo      | A       |
++----------+---------+
+| foo.v1   | A       |
++----------+---------+
+
+(2) PUT foo (second put), versioning: ``true``
+
++----------+---------+
+| key      | value   |
++==========+=========+
+| foo      | B       |
++----------+---------+
+| foo.v2   | B       |
++----------+---------+
+| foo.v1   | A       |
++----------+---------+
+
+Null Version Management
+^^^^^^^^^^^^^^^^^^^^^^^
+
+In a bucket without versioning, or when versioning is suspended, putting
+an object with the same name twice should result in the previous object
+being overwritten. This is managed with null versions.
+
+Only one null version should exist at any given time, and it is
+identified in S3 requests and responses with the version id "null".
+
+Case 1: Putting Null Versions
+'''''''''''''''''''''''''''''
+
+With respect to metadata, since the null version is overwritten by
+subsequent null versions, the null version is initially stored in the
+master key alone, as opposed to being stored in the master key and a new
+version. S3 checks if versioning is suspended or has never been
+configured, and sets the ``versionId`` option to ``''`` in PUT calls to
+the metadata engine when creating a new null version.
+
+The tables below show the keys resulting from PUT calls to metadata if
+we put an object 'foo' twice, when versioning has not been enabled or is
+suspended.
+
+(1) PUT foo (first put), versionId: ``''``
+
++--------------+---------+
+| key          | value   |
++==============+=========+
+| foo (null)   | A       |
++--------------+---------+
+
+(2) PUT foo (second put), versionId: ``''``
+
++--------------+---------+
+| key          | value   |
++==============+=========+
+| foo (null)   | B       |
++--------------+---------+
+
+The S3 API also sets the ``isNull`` attribute to ``true`` in the version
+metadata before storing the metadata for these null versions.
+
+Case 2: Preserving Existing Null Versions in Versioning-Enabled Bucket
+''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+Null versions are preserved when versioning has been enabled or
+re-enabled, and new non-null versions are created.
```

Comment:
I can edit this myself when I update the document, or you can, but I think this is clearer:
Null versions are preserved when new non-null versions are created after versioning has been enabled or
re-enabled.